Orange/widgets/unsupervised/owtsne.py

```diff
@@ -248,154 +431,159 @@ def set_data(self, data: Table):
+        assert task.pca_components == self.pca_components
+        assert isinstance(task.pca_projection, Table) and \
+            len(task.pca_projection) == len(self.data)
+
+    def __ensure_task_same_for_initialization(self, task: Task):
+        assert isinstance(task.initialization, np.ndarray) and \
+            len(task.initialization) == len(self.data)
+
+    def __ensure_task_same_for_affinities(self, task: Task):
+        assert task.perplexity == self.perplexity
+        assert task.multiscale == self.multiscale
+
+    def __ensure_task_same_for_embedding(self, task: Task):
+        assert task.exaggeration == self.exaggeration
+        assert isinstance(task.tsne_embedding, manifold.TSNEModel) and \
+            len(task.tsne_embedding.embedding) == len(self.data)
+
+    def on_partial_result(self, value):
+        # type: (Tuple[str, Task]) -> None
+        which, task = value
+
+        if which == "pca_projection":
+            self.__ensure_task_same_for_pca(task)
+            self.pca_projection = task.pca_projection
+        elif which == "initialization":
+            self.__ensure_task_same_for_pca(task)
+            self.__ensure_task_same_for_initialization(task)
+            self.initialization = task.initialization
+        elif which == "affinities":
+            self.__ensure_task_same_for_pca(task)
+            self.__ensure_task_same_for_affinities(task)
+            self.affinities = task.affinities
+        elif which == "tsne_embedding":
+            self.__ensure_task_same_for_pca(task)
+            self.__ensure_task_same_for_initialization(task)
+            self.__ensure_task_same_for_affinities(task)
+            self.__ensure_task_same_for_embedding(task)
+
+            prev_embedding, self.tsne_embedding = self.tsne_embedding, task.tsne_embedding
+            self.iterations_done = task.iterations_done
+            # If this is the first partial result we've gotten, we've got to
+            # setup the plot
+            if prev_embedding is None:
+                self.setup_plot()
+            # Otherwise, just update the point positions
+            else:
                 self.graph.update_coordinates()
                 self.graph.update_density()
-                self.needs_to_draw = False
-            # schedule next update
-            self.__timer.start()
-
-        self.__in_next_step = False
+        else:
+            raise RuntimeError(
+                "Unrecognized partial result called with `%s`" % which
+            )
 
-    def setup_plot(self):
-        super().setup_plot()
-        self.start()
+    def on_done(self, task):
+        # type: (Task) -> None
+        self.run_button.setText("Start")
+        # NOTE: All of these have already been set by on_partial_result,
+        # we double check that they are aliases
+        if task.pca_projection is not None:
+            self.__ensure_task_same_for_pca(task)
+            assert task.pca_projection is self.pca_projection
```

Comment:
Since you already have a method with assert statements (self.__ensure_task_same_for_pca()), why not putting this line there? You'd only have to make sure that the method is called after setting `self.pca_projection` in `self.on_partial_result()`.